src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +72,277 @@ static bool supervisor_restart_service(Supervisor *supervisor,
+}
+
+
+#define foreach_supervised_service(it, supervisor, service) \
+	for ((service) = supervisor_it_first((it), (supervisor)); \
+		 (service) != NULL; \
+		 (service) = supervisor_it_next(it))
+
+
+/*
+ * supervisor_add_dynamic starts the service and if successfull, adds it to the
+ * dynamicService array.
+ *
+ * Since the services are largely discoverable by name, the name has to be
+ * unique. If it is not, then the service is not started and the function
+ * returns false.
+ * Also since the service runs under supervision, there must be space for it to
+ * be added to the supervisor's dynamic services array. If there is not, then
+ * the service is not started and the function returns false.
+ * The service will not be added if a signal other than "dynamic service" has
+ * been received.
+ *
+ * Returns true if successfully starts the service.
+ */
+bool
+supervisor_add_dynamic(Supervisor *supervisor, Service *service)
+{
+	Service *dynamicService;
+	sigset_t sig_mask;
+	sigset_t sig_mask_orig;
+	int serviceIndex = supervisor->dynamicServices.serviceCount;
+	bool started;
+
+	if (supervisor_service_exists(supervisor, service->name))
+	{
+		log_error("Service %s already exists under supervision",
+				  service->name);
+		return false;
+	}
+
+	/* 0-based */
+	if (serviceIndex >= (MAXDYNSERVICES - 1))
+	{
+		log_error("Reached maximum permitted number of dynamic services, "
+				  "service %s is not added",
+				  service->name);
+		return false;
+	}
+
+	/*
+	 * Block signals until the service has been added in the dynamic array
+	 * unless failed to start. Failing to block signals, opens a race window.
+	 */
+	if (!block_signals(&sig_mask, &sig_mask_orig))
+	{
+		return false;
+	}
+
+	/*
+	 * In case there is an active signal to stop, set prior to blocking signals,
+	 * do not add the service. Our caller should be able to handle this case.
+	 */
+	if (asked_to_stop || asked_to_stop_fast || asked_to_reload || asked_to_quit)
+	{
+		/* restore signal masks (un block them) now */
+		(void) unblock_signals(&sig_mask_orig);
+		return false;
+	}
+
+	started = (*service->startFunction)(service->context, &(service->pid));
+
+	if (started)
+	{
+		uint64_t now = time(NULL);
+		RestartCounters *counters = &(service->restartCounters);
+
+		counters->count = 1;
+		counters->position = 0;
+		counters->startTime[counters->position] = now;
+
+		log_info("Started pg_autoctl %s service with pid %d",
+				 service->name, service->pid);
+	}
+	else
+	{
+		log_error("Failed to start pg_autoctl %s service",
+				  service->name);
+		(void) unblock_signals(&sig_mask_orig);
+		return false;
+	}
+
+	/*
+	 * IGNORE-BANNED fits here because we want to take ownership of the service.
+	 * We copy to an address that we own from a non overlapping address.
+	 */
+	dynamicService = &supervisor->dynamicServices.services[serviceIndex];
+	memcpy(dynamicService, service, sizeof(*service)); /* IGNORE-BANNED */
+	supervisor->dynamicServices.serviceCount++;
```

Comment:
Is there a way to just copy the structure instance directly, knowing that there is no pointer in the structure, it's all direct memory usage?

Reply:
Please if you have a specific function or call in mind, I will happily add it. I guess one can bend strcpy and friends accordingly but in essence it is quite the same. Those bytes from that address should be copied to this address.

Anything you want is fine, just point me to it please :)

Reply:
When working with struct instances (and more so when the structure internal slots are all “direct”, we are not dealing with pointers to other parts of the memory here, it is self-contained), then just use the `=` assignment operator.